File: python/21-30/Reverse Nodes in k-Group.py
class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        if k <=1 or head==None :
            return head
        su = ListNode(0)
        su.next=head
        pre,p = su,head
        while True:
            counter=0
            while p!=None and counter < k:
                p=p.next
                counter+=1
            if counter!=k:
                break
            #p is the next k-group begin
            t = pre.next
            curpre =pre.next
            cur=curpre.next
            while cur!=p:
                curpre.next =cur.next
                pre.next=cur
                cur.next=t
                #update
                t = cur
                cur=curpre.next
            pre = curpre
        return su.next

File: Cpp/21-30/Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
 /*
    runtime: 24ms
    error:1  
        if(head!=NULL)   //forget this condition
            temp->next = beg;
      
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* thead = new ListNode(0);
        ListNode* temp;
        while(head!=NULL){
            temp = head->next;
            head->next = thead->next;
            thead->next = head;
            head = temp;
        }
        return thead->next;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(k<2) return head;
        ListNode* thead= new ListNode(0);
        ListNode* temp = thead;
        ListNode* beg;
        while(head!=NULL){
          int i=0; 
          beg = head;
          while(i<k-1&&head->next!=NULL){
            i++;
            head =head->next;
          }
          if(i!=k-1) break;
          ListNode* t = head->next;
          head->next = NULL;
          temp->next = reverseList(beg);
          temp = beg;
          head =t;
        }
        if(head!=NULL)
            temp->next = beg;
        return thead->next;
    }
};
